dpath/segments.py

```diff
@@ -22,9 +25,11 @@ def leaf(thing):
 
     leaf(thing) -> bool
     '''
-    leaves = (bytes, str, int, float, bool, type(None))
-
-    return isinstance(thing, leaves)
+    return (
+        not isinstance(thing, collections.abc.Container)
```

Comment:
This will break on Python 2. At the top, put:

```py
try:
    import collections.abc as collections_abc
except ImportError:
    # Python 2
    import collections as collections_abc
```

This line would then use `collections_abc.Container` instead of `collections.abc.Container`

Reply:
v2 and onward will not support python 2.7 because it is EOL.

Reply:
Also, `isinstance("a string", collections.abc.Container)` so we can't do this directly without more logic.

Reply:
Oh, I see. I saw some other compatibility stuff so I assumed you were still supporting it.

I recommend adding the `python_requires` directive into your setup.py file to prevent it from installing on Python 2. Without it, users will be able to install it but it'll break with weird errors if they hit certain code paths.

Reply:
i get recursion error with only **return not isinstance(thing, collections.abc.Container).**

For me , it works with:
   **return isinstance(thing, (bytes, str, int, float, bool, type(None))) or not isinstance(thing, collections.abc.Container)**

best regards

Reply:
Weird...what was the data you were working with?

Reply:
Yes, it's weird.

 # Test with dates
    test = {
        "z": 29.9384,
        "a": {
            "b": {
                "3": datetime.date.today(),
                "43": 30.18,
                "c": [],
                "d": ['red', 'buggy', 'bumpers'],
            }
        }
    }
    dpath.util.get(test, '/a/b/3')
    dpath.util.set(test,'/a/b/3',31.18)

- collections.abc.Container raise a recursion exception. I'm using  python 3.7.7 on windows.
![imagen](https://user-images.githubusercontent.com/68324385/87632109-fd4d0280-c738-11ea-99fb-97b22d6958a9.png)

**Exception has occurred: RecursionError
maximum recursion depth exceeded in comparison**
  File "C:\programas\python\Python37\Lib\abc.py", line 139, in __instancecheck__
    return _abc_instancecheck(cls, instance)
  File "D:\Proyectos\datalakehouse\implementacion\python\shared\dpath\segments.py", line 25, in leaf
    return not isinstance(thing, collections.abc.Container)
  File "D:\Proyectos\datalakehouse\implementacion\python\shared\dpath\segments.py", line 44, in walk
    if not leaf(obj):
  File "D:\Proyectos\datalakehouse\implementacion\python\shared\dpath\segments.py", line 59, in walk
    for found in walk(v, location + (k,)):
  File "D:\Proyectos\datalakehouse\implementacion\python\shared\dpath\segments.py", line 59, in walk
    for found in walk(v, location + (k,)):
  File "D:\Proyectos\datalakehouse\implementacion\python\shared\dpath\segments.py", line 59, in walk
    for found in walk(v, location + (k,)):
  [Previous line repeated 975 more times]
  File "D:\Proyectos\datalakehouse\implementacion\python\shared\dpath\segments.py", line 346, in fold
    for pair in walk(obj):
  File "D:\Proyectos\datalakehouse\implementacion\python\shared\dpath\util.py", line 172, in get
    results = dpath.segments.fold(obj, f, [])
  File "D:\Proyectos\datalakehouse\implementacion\python\test\test_transform.py", line 30, in main3
    dpath.util.get(test, '/a/b/3')


My leaf function:
![imagen](https://user-images.githubusercontent.com/68324385/87632408-a562cb80-c739-11ea-9f75-69755f46622e.png)

Hope it help.
Regards

Reply:
dpath version:
VERSION = "2.0.1"

Reply:
Oh wait you're missing the
```py
    or isinstance(thing, six.string_types)
    or isinstance(thing, six.binary_type)
)
```
part